lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -4282,3 +4282,81 @@ AnyFunctionType *swift::adjustFunctionTypeForConcurrency(
 bool swift::completionContextUsesConcurrencyFeatures(const DeclContext *dc) {
   return contextUsesConcurrencyFeatures(dc);
 }
+
+/// This is a best-effort to ensure that you aren't directly calling API that is
+/// annotated with an @_unavailableFromAsync attribute.
+/// This is fairly easy to get around. Simply wrapping a call in a closure
+/// will allow calling the API from the synchronous context
+void swift::checkAsyncAvailability(AbstractFunctionDecl &decl) {
+  if (decl.hasAsync() && decl.getAttrs().hasAttribute<UnavailableFromAsyncAttr>())
+    decl.getASTContext().Diags.diagnose(decl.getLoc(),
+        diag::warn_async_function_must_be_available_from_async,
+        decl.getName());
```

Comment:
Presumably we need the same kind of check for an unavailable-from-async property, subscript, or initializer, all of which can be `async`. Could you move this to `TypeCheckAttr.cpp`, where we validate the correctness of attributes that are applied to declarations?